src/lib.rs

```diff
@@ -138,7 +138,7 @@ macro_rules! imap_deconstruct {
 mod test {
     use super::*;
 
-    fn host_library<T: ImplicitClone>(value: &T) -> T {
```

Comment:
why? the point of the test wasn't just testing, it was an example of the "host library" term used in documentation

Reply:
well if it's in test, the point is to test, and right now I need the test to show to the developer I am actually adding impls on types that are actually Copy

Reply:
oh wait, the function is called "host_library" hahaha you're right

Reply:
> well if it's in test, the point is to test

wrong, the point of a test is to document code in a way that has stronger guarantees (always actual info) than a comment or some sort of a separate wiki

Reply:
I fixed it but you won't like it xD if you want doc you put it in doc though. I think that makes more sense. Who's gonna check the crate's test to see how a crate is used anyway :woman_facepalming:

Reply:
> Who's gonna check the crate's test to see how a crate is used anyway

me, i know what tests are supposed to be used for (at least 50% of people i hear mention "test" and "documentation" in same sentence even when they are not talking about doctests) so they are valuable to me as a user, especially if some author's docs are lacking but their testing is somewhat diligent

> you won't like it

eh just comment i guess that "hey this `assert_implicit_clone!` could be your example of a host library"

Reply:
Rust has a fantastic tooling for documentation:
- rustdoc
- doc tests (which are both tests and documentation, that's where you should put that)
- examples
- ...

And you want to put part of the documentation into the internal tests. I understand, I would have done that in JS, Python, ... but absolutely not in Rust. If Rust is that good it's also because of it's tooling around the language itself.

I don't mind moving some tests to doc tests if you want.

Reply:
> Python

Python is the origin of doctests lol
Many Rust features have come from Python

> I don't mind moving some tests to doc tests if you want

If its possible, find a spot where to show to the user that they don't get implicit cloning in their code unless they make a host library (a macro)

Reply:
> If its possible, find a spot where to show to the user that they don't get implicit cloning in their code unless they make a host library (a macro)

I'm trying to find an example but the only thing I can think of is Yew's html itself. Maybe I can write an example with a simplified `html!` macro. what do you think?

Reply:
> wrong, the point of a test is to document code in a way that has stronger guarantees (always actual info) than a comment or some sort of a separate wiki

I mean the word itself is "test", at no point it's mentioning "document code". It's a single word and it's basic meaning is to test. It's true that I also sometimes consult the tests of a library to see how to use it but that's more because the thing is poorly documented. That never happened to me with crates in Rust, I never read Yew's tests to understand how to use the thing.

Reply:
all you need to show is that you no longer require clones

```rust
let a = SomeImmovableType;

host_library(a);
host_library(a); // fails, previous line moved the value

host_library!(a);
host_library!(a); // works, previous line cloned implicitly
```

Reply:
Maybe I should go with something simpler like you show here. My example isn't super good :(

Reply:
As you can see [here](https://github.com/yewstack/implicit-clone/issues/43) and [here](https://github.com/yewstack/implicit-clone/commit/ddb351ce556972d5edeaf0c0e7bdd54b033b1d83#commitcomment-131770601) it seems that people are still not understanding the point this crate despite our efforts to clarify it in #28.

What's interesting here is that you are right, people are definitely looking at the tests to see how it works! Though, I'm still in favor of providing a better and more sensical example within the documentation directly (crate doc + README).

Besides, I still would like the test to assert that we don't implement ImplicitClone on types that are not Copy.

(@kirillsemyonkin please bear with me as I really want to get to the bottom of this. As long as people are opening issues because they misunderstand or don't see how to use the crate, it means that there is work to do there and we should try our best to clarify things out.)